src/com/palmergames/bukkit/towny/tasks/DailyTimerTask.java

```diff
@@ -206,23 +201,69 @@ protected void collectNationTaxes(Nation nation) throws EconomyException {
 				 * exists.
 				 * We are running in an Async thread so MUST verify all objects.
 				 */
-				if (townyUniverse.getDataSource().hasTown(town.getName())) {
+				if (universe.getDataSource().hasTown(town.getName())) {
 					if (town.isCapital() || !town.hasUpkeep())
 						continue;
-					if (!town.getAccount().payTo(nation.getTaxes(), nation, "Nation Tax")) {
-						localRemovedTowns.add(town.getName());		
-						town.removeNation();
+					if (town.getAccount().canPayFromHoldings(taxAmount)) {
+					// Town is able to pay the nation's tax.
+						town.getAccount().payTo(taxAmount, nation, "Nation Tax to " + nation.getName());
+						TownyMessaging.sendPrefixedTownMessage(town, Translation.of("msg_payed_nation_tax", TownyEconomyHandler.getFormattedBalance(taxAmount)));
 					} else {
-						TownyMessaging.sendPrefixedTownMessage(town, TownySettings.getPayedTownTaxMsg() + nation.getTaxes());
+					// Town is unable to pay the nation's tax.
+						if (!TownySettings.isTownBankruptcyEnabled() || (TownySettings.isTownBankruptcyEnabled() && !TownySettings.doBankruptTownsPayNationTax())) {
+						// Bankruptcy disabled, remove town for not paying nation tax, 
+						// OR Bankruptcy enabled but towns aren't allowed to use debt to pay nation tax. 
+							localNewlyDelinquentTowns.add(town.getName());		
+							town.removeNation();
+							TownyMessaging.sendPrefixedTownMessage(town, Translation.of("msg_your_town_couldnt_pay_the_nation_tax_of", TownyEconomyHandler.getFormattedBalance(taxAmount)));
+							continue;
+						}
+
+						// Bankruptcy enabled and towns are allowed to use debt to pay nation tax.
+						boolean townWasBankrupt = town.isBankrupt();
+						town.getAccount().setDebtCap(MoneyUtil.getEstimatedValueOfTown(town));
+						
+						if (town.getAccount().getHoldingBalance() - taxAmount < town.getAccount().getDebtCap() * -1) {
+						// Towns that would go over their debtcap to pay nation tax, need the amount they pay reduced to what their debt cap can cover.
+						// This will result in towns that become fully indebted paying 0 nation tax eventually.
+							taxAmount = town.getAccount().getDebtCap() - Math.abs(town.getAccount().getHoldingBalance());

```

Comment:
I would put this line after the if-statement below since it's not used in that statement below.

Reply:
Doing the same to the CollectTownCosts saves creating the trueUpkeep variable as well.